service/frontend/dcRedirectionHandler.go

```diff
@@ -30,23 +35,59 @@ import (
 type (
 	// DCRedirectionHandlerImpl is simple wrapper over fontend service, doing redirection based on policy
 	DCRedirectionHandlerImpl struct {
-		currentClusteName string
-		redirectionPolicy DCRedirectionPolicy
-		tokenSerializer   common.TaskTokenSerializer
-		sevice            service.Service
-		frontendHandler   *WorkflowHandler
+		currentClusterName string
+		domainCache        cache.DomainCache
+		config             *Config
+		redirectionPolicy  DCRedirectionPolicy
+		tokenSerializer    common.TaskTokenSerializer
+		service            service.Service
+		frontendHandler    *WorkflowHandler
 	}
 )
 
+const (
+	// DomainActivenessAttemptCount indicate the attempts for the following API
+	// if seeing DomainNotActiveError
+	//
+	// 1. StartWorkflowExecution
+	// 2. SignalWithStartWorkflowExecution
+	// 3. SignalWorkflowExecution
+	DomainActivenessAttemptCount = 2
+
+	// APIAttemptCount is the attempt for normal API calls
+	APIAttemptCount = 1
+)
+
+// NewDCRedirectionHandlerWithWorkflowHandler create a dc redirection handler with existing frontend handler
+func NewDCRedirectionHandlerWithWorkflowHandler(wfHandler *WorkflowHandler, policy config.DCRedirectionPolicy) *DCRedirectionHandlerImpl {
+	dcRedirectionPolicy := RedirectionPolicyGenerator(
+		wfHandler.GetClusterMetadata(),
+		wfHandler.domainCache,
+		policy,
+	)
+	dcRedirectionHandler := NewDCRedirectionHandler(
+		wfHandler.GetClusterMetadata().GetCurrentClusterName(),
+		wfHandler.domainCache,
+		wfHandler.config,
+		dcRedirectionPolicy,
+		wfHandler.Service,
+		wfHandler,
+	)
+
+	return dcRedirectionHandler
+}
+
 // NewDCRedirectionHandler creates a thrift handler for the cadence service, frontend and admin
-func NewDCRedirectionHandler(currentClusteName string, redirectionPolicy DCRedirectionPolicy,
-	sevice service.Service, frontendHandler *WorkflowHandler) *DCRedirectionHandlerImpl {
+func NewDCRedirectionHandler(currentClusterName string, domainCache cache.DomainCache, config *Config, redirectionPolicy DCRedirectionPolicy,
```

Comment:
this method also takes a WorkflowHandler as input i.e. looks like most of the other inputs this constructor takes could also be derived from workflowHandler - only difference between this and other constructor i see is - this method takes a policy directly whereas other takes a redirectPolicyConfig - is this correct ? If yes, can you fix this ?